packages/core/src/parser/internal/parse.ts

```
@@ -204,31 +217,40 @@ export const parseBuffer = (
       ].join('')
       setErrorRecoveryMode() // Allows the wildcard token to be parsed from now on in this file
 
-      const [finalFixedBuffer, blockItems, errors] = parseBuffer(
+      const [finalFixedBuffer, elements, sourceMap, errors] = await parseBuffer(
         fixedBuffer,
         filename,
+        functions,
         [...prevErrors, parserError],
         true
       )
       if (!errorRecoveryMode) {
         log.debug('Finished parsing: %s (%d bytes)', filename, errors.length)
       }
       return [finalFixedBuffer,
-        restoreOrigBlockRanges(blockItems, parserError.subject.start), errors]
+        elements,
+        restoreOrigBlockRanges(sourceMap, parserError.subject.start),
+        errors,
+      ]
     }
-    return [src, [], [...prevErrors, parserError]]
+    return [src, [], new SourceMap(), [...prevErrors, parserError]]
   }
   if (!errorRecoveryMode) {
     log.debug('Finished parsing: %s', filename)
   }
-  const blockItems = hclParser.finish()[0]
-  return [src, blockItems, prevErrors]
+  const parseResult = hclParser.finish()[0]
+  await replaceFunctionValues()
```

two questions:
1. Is there a point in doing this if the `parseResult` is undefined?
2. IIUC, if one of the promises that are found throws an error, this will basically fail everything and mask all other errors. is that the behavior we want?

also, as mentioned before, this seems to have nothing to do with functions

---

1. It clears the array that holds the (a)waiting promises 
2. I kind of think that is what we want here. What ever creates the promises should wrap itself in a try catch and return some kind of meaningful value instead of the failed promise. (For example, the static file function returns `InvalidStaticFile`). -  Thoughts?

---

Given all the effort you made in trying to make the parser recover and provide meaningful errors, it just seems strange that you'd throw all that away if one function (that we get from outside the parser) misbehaves.
The error itself won't be useful as well because you won't get any information about where it happened in the files...
anyway, I don't think it is wise to rely on functions not throwing but up to you